### DISCUSSION TITLE
CI Build Proposal

### DISCUSSION THREAD

**Discussion by sapphire-arches** [SynthstromAudible/DelugeFirmware#65] [Ideas]
For all the usual reasons we probably want the firmware builds to have some amount of CI coverage. I'll have some time to implement this but other contributions are of course welcome. My current plan is as follows:

# Build Image
To hopefully minimize build time, I think we should create a docker image that has the [dbt-toolchain](https://github.com/litui/dbt-toolchain) pre-installed. We should base it on something small because github actions re-downloads the image on every build, but not so small that it gets complicated to get the toolchain running. This likely means `debian:latest` or `ubuntu:latest` from docker. The images can be hosted on GHCR, with actions (in the toolchain repo?) that publish the container automatically.

# Build Matrix

The current build matrix looks like:
- Build Type
  - Debug
  - Release
- Hardware Platform
  - OLED
  - 7SEG

It's possible in the future we will merge the OLED and 7SEG builds (turning it in to a runtime switch).

**Comment by PaulFreund** [comment]
I have to read up a little bit on what GitHub provides here (my experience is mainly with Jenkins) but the proposal sounds solid. I'd like to vote for ```debian:latest```. If you have tasks to assign hit me up!

**Comment by sapphire-arches** [comment]
A prototype of the images for this is here: https://github.com/bobtwinkles/deluge-ci-images/blob/main/Dockerfile.

I ended up using `ubuntu:latest` because it's actually a smaller base image, though the ~10mb difference disappears as a rounding error in the face of the ~1GB for the toolchain.

@litui, when you have a moment I'd appreciate a review of the Dockerfile I've kludged together here: https://github.com/bobtwinkles/deluge-ci-images/blob/55cb97034f7fe72cb4f78a9e4bb5fee36c601008/Dockerfile. I think I'm plugging in to DBT at the right abstraction points, but there's at least a little bit of weirdness with how I handle getting the wheels (pre-)installed. The goal of that step is 1) reduce final image size and 2) remove the need for the toolchain directory to be r+w in the final image which seemed to cause some issues (docker liked to assign a weird UID/GID).

**Reply by litui** [reply]
Hmm... I like what you're doing here. Deleting the wheels isn't a bad idea (I should probably do that in ./dbt so it doesn't retry on every run), but the toolchain really should be read/write for python at least. I'd like to reserve the option to remotely install new dependencies through ./dbt without repackaging everything, especially for wheels that are platform-specific. I'd suggest creating the /dbt path and chowning it to your nonroot user, then switching to nonroot before running dbt.

Also, I'd recommend reading the Docker "Architecture" setting to know which linux toolchain to use. For instance, if I tried making a Docker image from this on M2 mac it'd probably fail out of the box. This would make it compatible with xbuild for multiarch packaging.

**Reply by litui** [reply]
I'd also like to make ./dbt outside the image work with your docker image *if present* by having it check for docker and an image with the right label. Then ./dbt will continue to "just work" even in your case ;)

**Reply by sapphire-arches** [reply]
I think I've addressed both of these concerns. The latest version:
- Uses `TARGETARCH` to determine what DBT to package. The Actions workflow builds both amd64 (x86_64) and aarch64 (arm64) images now.
- The python toolchain directory should be read/write now.

For posterity, the incantation to actually test these toolchains is:

```shell
docker run -v $(pwd):/src --rm ghcr.io/bobtwinkles/deluge-ci-images:main
```

This uses the `dbt` scripts from inside the current directory but the toolchain itself is baked in to the image. The images themselves are around 300-350 megabytes in size.

**Comment by sapphire-arches** [comment]
There's a real PR up here now: https://github.com/SynthstromAudible/DelugeFirmware/pull/72
